Make Labels Optional for CounterMetric::add

astorm opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

This solves the problem described in #1031 -- where not providing labels to a counter metric results in an error. 

## Short description of the changes

This PR changes the method signature of the `add` method on the `CounterMetric` class to provide a default value for the `labels` parameter.  This is done so that [the `hashLabels` utility function](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-metrics/src/Utils.ts#L33) does not throw an error when trying to pass `undefined` to `Object.keys(labels);`

This PR also adds a new unit test to exercise the default option.

mayurkale22 commented on packages/opentelemetry-metrics/src/Metric.ts:
```diff
@@ -133,7 +133,7 @@ export class CounterMetric extends Metric<BoundCounter> implements api.Counter {
    *     that you want to record.
    */
-  add(value: number, labels: api.Labels) {
+  add(value: number, labels: api.Labels = {}) {
```
This is nice, it would be nice if you can do the same thing to `measure.record` (https://github.com/astorm/opentelemetry-js/blob/astorm/labels-optional/packages/opentelemetry-metrics/src/Metric.ts#L166)

mayurkale22 replied:
Just realized, Prometheus exporter would fail after this change. Especially below line due to `toString` method on `undefined`:

https://github.com/open-telemetry/opentelemetry-js/blob/c454d6cf2c895be39e7ae9d1b55c039ed952dcb4/packages/opentelemetry-exporter-prometheus/src/prometheus.ts#L187

I would prefer to fix it in this same PR and do some end to end manual testing. WDYT?

mayurkale22 replied:
Also, in order to pass the build you need to make the same change in the API package.

astorm replied:
Thanks for the guidance and patience @mayurkale22 -- I'm new to the repo and the project and I'm still getting my sea legs.  

Fixing this for the `MeasureMetric` in this PR makes sense.  Fixing the prometheus exporter in this PR also makes sense.  I presume the change in `opentelemetry-api` would be to the interfaces, and that also makes sense to do in this PR. 

I'll ping -- you? -- via a PR comment when that's done.  If there's more to The Process™ please let me know. 

Best wishes.

mayurkale22 replied:
SGTM, just add a comment on this thread when new changes are ready to review. I usually use [this example](https://github.com/open-telemetry/opentelemetry-js/tree/master/examples/prometheus) to test against  prom. exporter.

dyladan replied:
@mayurkale22 not sure why you think the prometheus exporter would fail on that line. Can you explain? I'm hesitant to approve changes to the `_registerMetric` routine in the prometheus exporter because the prom-client has some surprising behavior, and that is the logic we use to destroy and recreate the counters with the new values.

mayurkale22 replied:
> @mayurkale22 not sure why you think the prometheus exporter would fail on that line. Can you explain?

Looks like @astorm already mention the reason, here-> https://github.com/open-telemetry/opentelemetry-js/pull/1032#discussion_r421580110 (due to `toString` method on `undefined` label value).

> I'm hesitant to approve changes to the _registerMetric routine in the prometheus exporter because the prom-client ....

I would suggest to add new test in prom. exporter with some `LabelKeys` but w/o `Labels` (values). You can reuse existing test ([like this one](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-exporter-prometheus/test/prometheus.test.ts#L200)), just dont bind the labels to metric.

astorm commented:
@mayurkale22 (or others) -- I've

- updated the `MeasureMetric.record` signature so labels is optional with a default value of `{}` -- existing test coverage seems sufficient here, so I did not add an additional unit test.

- updated `interface Counter` and `interface Measure` to make labels an optional parameter there as well

Regarding Prometheus -- [I left a question in the gitter room](https://gitter.im/open-telemetry/opentelemetry-node?at=5eb3583414b48f0698b40060) about this, but I can't get the prometheus exporter package to build cleanly.  When running `npm install` from the root, I end up with [the following results](https://gist.github.com/astorm/b77d88f29c55f5fb7717e87af0df5b76).  TLDR -- is _seems_ like the prometheus package at `master` can't compile right now. 

    test/prometheus.test.ts:527:5 - error TS7027: Unreachable code detected.

    527     done();
            ~~~~~~~


    Found 2 errors.

    npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! @opentelemetry/exporter-prometheus@0.7.0 compile: `npm run version:update && tsc -p .`
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the @opentelemetry/exporter-prometheus@0.7.0 compile script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

    npm ERR! A complete log of this run can be found in:
    npm ERR!     /Users/alanstorm/.npm/_logs/2020-05-07T03_59_24_975Z-debug.log
    lerna info lifecycle @opentelemetry/exporter-collector@0.7.0~prepare: @opentelemetry/exporter-collector@0.7.0
    lerna info lifecycle @opentelemetry/exporter-jaeger@0.7.0~prepare: @opentelemetry/exporter-jaeger@0.7.0
    lerna info lifecycle @opentelemetry/exporter-zipkin@0.7.0~prepare: @opentelemetry/exporter-zipkin@0.7.0
    lerna info lifecycle @opentelemetry/node@0.7.0~prepare: @opentelemetry/node@0.7.0
    lerna info lifecycle @opentelemetry/shim-opentracing@0.7.0~prepare: @opentelemetry/shim-opentracing@0.7.0
    lerna info lifecycle @opentelemetry/test-utils@0.7.0~prepare: @opentelemetry/test-utils@0.7.0
    lerna info lifecycle @opentelemetry/web@0.7.0~prepare: @opentelemetry/web@0.7.0
    lerna info lifecycle @opentelemetry/exporter-prometheus@0.7.0~prepare: Failed to exec prepare script.  

Before I go off on some yak shave, does anyone reading have any advice on how I might proceed to a clean build.  One I have a clean build I can make a pass at fixing the prometheus exporter so it doesn't assume counter metrics have labels.

mayurkale22 commented:
> test/prometheus.test.ts:527:5 - error TS7027: Unreachable code detected.

I tried repro'd it locally, seems working fine for me.

astorm commented:
@mayurkale22 OK -- I was able to work around (but not fix) my repo/learna issues and fix the Prometheus exporter for the case where A metric's been defined with set of label keys, but values are not provided when recording the metric (either by passing a `{}` to `add`, or by using this new default value and passing nothing to `add`)

My main interest in this PR is being able to create a simple counter with no labels and increment it.  As this PR sits right now, I think we've accomplished that, but if there's more to be done here please let me know.

Also -- just to say it out loud -- there seems to be a non-symmetry to the API design that exists regardless of this PR.  The fact you can define a set of label keys for a metric, but then _not_ define values for those keys when recording a metric creates some ambiguity as far as what the correct behavior should be (Record with _undefined_ values for that key(s)? Throw an error because the keys are there? _Log_ an error but still record? Log an error but don't record?)  I'm not sure if anything can/should be done about this, but if there is it feels outside the scope of this small PR.

Thanks again for considering this and for the help getting oriented in the repo.

dyladan commented:
@astorm since you're in a weird state I would suggest a complete clean and rebootstrap

MAKE SURE YOU HAVE COMMITTED ANYTHING YOU WANT TO KEEP BEFORE DOING THIS

```sh
# from the root of the project
lerna clean -y
git clean -fdx
npm install # This will cause the project to be built and linked by npm scripts
```

astorm commented:
@dyladan Thank you for the advice -- I gave that a try and [still ran similar issues](https://gist.github.com/astorm/d4ac39447475aec3386f2a08be3db3a4), and I tried a fresh `git clone` in a separate directory and ran into similar issues.  The fresh checkout makes me think this isn't a problem of the repo being in some bad state, but rather there's some you'all take for granted about your environments that I'm not aware of .

@dyladan question -- I noticed in the sample command you gave me that you're running `lerna` globally.  I don't have `lerna` installed globally and was relying on the version installed in the project.  Could you let me know what version of `lerna` you have installed globally so I can attempt to match your environment? (also -- what version of node are you using?)

dyladan commented:
Oh sorry yes I do have lerna installed globally. The issue you're seeing looks like a compilation issue. It's possible that the `@types/node` package updated and you just happened to be the unlucky person to be affected first.

```sh
$ lerna --version
3.20.2
$ node --version
v12.14.1
```

dyladan commented:
I am able to reproduce your issue on a complete rebuild.

astorm commented:
> I am able to reproduce your issue on a complete rebuild. 

Yay, I am not any more insane that any other software engineer :)  And I learned the project's CircleCI tests don't include a full compile cycle, so that's a good thing to know.

Regardless of these build issues, this PR is ready for review.

dyladan commented:
The CI uses a cached set of dependencies because the dep install takes so long and takes so many resources. It's unfortunate, but it is the way it is. Changing any `package.json` causes the cache key to change and a full build in CI.

dyladan commented on packages/opentelemetry-exporter-prometheus/src/prometheus.ts:
```diff
@@ -184,7 +184,9 @@ export class PrometheusExporter implements MetricExporter {
     if (metric instanceof Counter) {
       metric.remove(
-        ...record.descriptor.labelKeys.map(k => record.labels[k].toString())
+        ...record.descriptor.labelKeys.map(k =>
```
In what cases would `record.labels[k]` be missing?

astorm replied:
In my manual testing last night this change prevented an error being thrown when a user instantiates a counter with a label key

    const monotonicCounter = meter.createCounter('monotonic_counter', {
      monotonic: true,
      labelKeys: ['pid'],
      description: 'Example of a monotonic counter',
    });

but then fails to provide a value for that key.  

    monotonicCounter.add(1, {})
    monotonicCounter.add(1)    

@dyladan With my build issues sorted/understood I can (depending on how your conversation with @mayurkale22 goes) either add tests that make the use-case here clear or revert this.

dyladan replied:
Ah I understand. Instead of falling back to an empty string, I think I would prefer a filter on the list.

```typescript
      metric.remove(
        ...record.descriptor.labelKeys
          .filter(k => record.labels[k] != null)
          .map(k => record.labels[k].toString())
      );
```

mayurkale22 replied:
This is clever solution but feel like it won't work as expected.

AFAIK prom-client will throw "**Invalid number of arguments**" exception if length of keys != length of values. See: https://github.com/siimon/prom-client/blob/master/lib/util.js#L34

One option is to fall back on "**value missing** or **unknown**" as a default value when the corresponding value is not defined in any given key(s), similar to the [description field](https://github.com/open-telemetry/opentelemetry-js/blob/8a29221c08a90fd6fe85970210b4c3c8a3867de7/packages/opentelemetry-exporter-prometheus/src/prometheus.ts#L200). WDYT?

mayurkale22 replied:
Option 2: Instead of [blindly attaching](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-exporter-prometheus/src/prometheus.ts#L201) our `Labelkeys` to prom-client's `labelNames`, we can filter the keys based on valid value.

This is what python has implemented: https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-prometheus/src/opentelemetry/ext/prometheus/__init__.py#L146-L150

astorm replied:
Thank you for your feedback @mayurkale22

> WDYT?

IT I've been wrangled into an bug fix that is _just_ outside the scope of my original PR -- but I can live with that :)

Re: option 2 -- when you say 

> Option 2: Instead of blindly attaching our Labelkeys to prom-client's labelNames, we can filter the keys based on valid value.

My interpretation of this is it's a suggestion that, when it comes time to create the Prometheus metric, we look at the keys and values of the passed in Open Telemetry metric, remove any keys with "non-valid" values (undefined, null, -- others?) and create the Prometheus metric with this new, smaller, set of keys. 

If that's the suggestion -- I don't think this is the right course of action.  Label keys seem like a fundamental part of a metric.  That is -- a metric with the keys [`foo`, `bar`, `baz`] seems like a _different_ metric than one with the keys [`foo`, `bar`].  Removing keys doesn't seems like the right course of action. 

Give than, Option 1 seems like the better approach.  

> One option is to fall back on "value missing or unknown" as a default value when the corresponding value is not defined in any given key(s),

You didn't indicate whether you thought this value should be added when creating the Prometheus metric, or when _recording_ the Open Telemetry metric.  My assumption is the former, as having this extra work happen _every_ time a metric is recorded would be excessive. 

Given all that: I intend to change how Prometheus metrics are instantiated.  If a value isn't present for a specific key, I will use a constant string of `value missing` for its value.  I will add tests to cover this case.  I'll check the spec for what constitutes a valid values, and make judgment call if there are ambiguities there.

mayurkale22 replied:
@dyladan would like you know your thoughts on this? 

> IT I've been wrangled into an bug fix that is just outside the scope of my original PR

Perhaps we can merge this PR and address prometheus exporter issue in the separate PR (btw, this won't make master branch unstable but we need to apply fix before next release). If we decide to merge, I would suggest to revert prometheus related to changes from this PR.

dyladan replied:
Think I would rather see it as two separate PRs simply for the purpose of having better history later when we have to go spelunking through PRs to discover why decisions were made. Not a strong preference though. Agree the prom changes should be reverted if this is going to merge without a proper fix.

mayurkale22 commented:
@open-telemetry/javascript-approvers We need more reviews for this one!

obecny commented on packages/opentelemetry-metrics/test/Meter.test.ts:
```diff
@@ -92,6 +92,19 @@ describe('Meter', () => {
       );
     });
 
+    it('be able to call add with no labels', () => {
```
Nit: "`should` be able to ...."

astorm replied:
thank you -- this nit's been deloused

obecny commented on packages/opentelemetry-metrics/src/Metric.ts:
```diff
@@ -133,7 +133,7 @@ export class CounterMetric extends Metric<BoundCounter> implements api.Counter {
    * @param labels key-values pairs that are associated with a specific metric
```
the doc should probably look like this after the param is optional
`* @param [labels = {}] key-values pairs ........`

astorm replied:
Thank you.  This docblock's type has been changed per the your correction.
